File: 2020/cpp/day01.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <numeric>
#include <set>
#include <string>
#include <vector>

bool is_2020(const std::vector<uint32_t> &items) {
  if (std::accumulate(items.begin(), items.end(), 0) == 2020)
    return true;
  return false;
}

uint32_t run(std::vector<uint32_t> &items) {
  for (auto i : items)
    for (auto j : items) {
      if (i == j)
        continue;
      if (is_2020(std::vector<uint32_t>{i, j}))
        return i * j;
    }
  return 0;
}

void read_input(std::vector<uint32_t> &items) {
  std::ifstream in("../inputs/input01");
  std::string str;
  while (std::getline(in, str)) {
    if (str.find_first_not_of(' ') != std::string::npos)
      items.push_back(std::stoi(str));
  }
}

void part1() {
  std::vector<uint32_t> items;
  read_input(items);
  std::cout << run(items) << std::endl;
}

uint32_t run2(std::vector<uint32_t> &items) {
  for (auto i : items)
    for (auto j : items)
      for (auto k : items) {
        if (i == j || i == k || j == k)
          continue;
        if (is_2020(std::vector<uint32_t>{i, j, k}))
          return i * j * k;
      }
  return 0;
}

void part2() {
  std::vector<uint32_t> items;
  read_input(items);
  std::cout << run2(items) << std::endl;
}

void test_is_2020() {
  assert(is_2020(std::vector<uint32_t>{1721, 299}));
  assert(is_2020(std::vector<uint32_t>{1900, 120}));
  assert(!is_2020(std::vector<uint32_t>{0, 1900}));
}

int main(void) {
  test_is_2020();
  part1();
  part2();
  return 0;
}

File: 2020/rust-sln/src/day01.rs
use combinations::Combinations;

fn main() {
    println!("Part1 {}", part1());
    println!("Part2 {}", part2());
}

fn is_2020(combinations: &Vec<usize>) -> bool {
    if let Some(_) = combinations.iter().find(|x| **x == 0) {
        return false;
    }

    return combinations.iter().map(|x| *x).sum::<usize>() == 2020;
}

fn run(items: &Vec<usize>, size: usize) -> usize {
    for combination in Combinations::new(items.clone(), size) {
        if is_2020(&combination) {
            return combination.iter().fold(1, |acc, x| acc * *x);
        }
    }
    0
}

fn read_input() -> Vec<usize> {
    include_str!("../../inputs/input01")
        .lines()
        .flat_map(|line| line.parse::<usize>())
        .collect()
}

fn part1() -> usize {
    run(&read_input(), 2)
}

fn part2() -> usize {
    run(&read_input(), 3)
}

mod tests {
    use super::*;

    #[test]
    fn test_is_2020() {
        assert!(is_2020(&vec![1721, 299]));
        assert!(!is_2020(&vec![0, 1]));
        assert!(!is_2020(&vec![1900, 110]));
    }

    #[test]
    fn test_run() {
        assert_eq!(run(&vec![1, 1721, 299], 2), 514579);
        assert_eq!(run(&vec![0, 1721, 2, 299], 2), 514579);
    }
}
